### PR TITLE
Add safety checks in configuration (fixes #216)

### LINKED ISSUE #216: Add safety checks for review configuration
* Not twice the same source
* Not different sources towards the same collection
* etc.

### PR DESCRIPTION
**Pull request by leplatrem**
Fixes #216

### CODE REVIEW

File: kinto_signer/utils.py [hunk 1]
```diff
@@ -35,53 +58,68 @@ def parse_resources(raw_resources):
...
+            source = _get_resource(source_uri)
+            destination = _get_resource(destination_uri)
+            preview = _get_resource(preview_uri) if preview_uri else None
+        except ValueError as e:
+            raise ConfigurationError(error_msg % e)
+
+        # Raise if mix-up of per-bucket/specific collection.
+        mixed = source['collection'] is None and (
+            destination['collection'] is not None or
+            preview is not None and preview['collection'] is not None)
```
Comment (glasserc) [inline kinto_signer/utils.py]: Should this check equally that if `destination['collection']` is None, `source['collection']` should be None as well?

File: kinto_signer/utils.py [hunk 2]
```diff
@@ -35,53 +58,68 @@ def parse_resources(raw_resources):
...
+        if mixed:
+            raise ConfigurationError(error_msg % "cannot mix bucket and collection URIs")
+
+        # Repeated source/preview/prod.
```
Comment (glasserc) [inline kinto_signer/utils.py]: Should `prod` be `dest`?

File: kinto_signer/utils.py [hunk 3]
```diff
@@ -35,53 +58,68 @@ def parse_resources(raw_resources):
...
             'destination': destination,
         }
         if preview is not None:
-            resources[key]['preview'] = _get_resource(preview)
-        # XXX: raise if mix-up of per-bucket/specific collection
-        # XXX: raise if same bid/cid twice/thrice
+            resources[key]['preview'] = preview
+
+    # Raise if same bid/cid twice/thrice.
+    # Theorically we could support it, but since we never said it was possible
```
Comment (glasserc) [inline kinto_signer/utils.py]: µnit: `Theoretically`

File: kinto_signer/utils.py [hunk 4]
```diff
@@ -35,53 +58,68 @@ def parse_resources(raw_resources):
...
+    # and have no test at all for that, prefer safety.
+    sources = [tuple(r['source'].items()) for r in resources.values()]
+    destinations = [tuple(r['destination'].items()) for r in resources.values()]
+    previews = [tuple(r['preview'].items()) for r in resources.values()
+                if 'preview' in r]
+
+    if len(set(destinations)) != len(destinations):
+        raise ConfigurationError("Resources setting has repeated destination URI")
+    if len(set(previews)) != len(previews):
+        raise ConfigurationError("Resources setting has repeated preview URI")
```
Comment (glasserc) [inline kinto_signer/utils.py]: Why don't we check `sources` here too?

File: tests/test_utils.py [hunk 5]
```diff
@@ -150,3 +150,74 @@ def test_resources_can_be_defined_per_bucket(self):
...
+
+        raw_resources = "/buckets/stage;/buckets/preview;/buckets/preview"
+        with self.assertRaises(ConfigurationError):
+            utils.parse_resources(raw_resources)
+
+    def test_cannot_repeat_resources(self):
+        # Repeated source.
+        raw_resources = """
+        /buckets/stage;/buckets/preview;/buckets/prod
+        /buckets/stage;/buckets/preview;/buckets/prod
```
Comment (glasserc) [inline tests/test_utils.py]: This test is a bit strange, it seems to cover a bunch of repetitions.

File: tests/test_utils.py [hunk 6]
```diff
@@ -150,3 +150,74 @@ def test_resources_can_be_defined_per_bucket(self):
...
+        /buckets/bid1;/buckets/stage;/buckets/bid2
+        """
+        with self.assertRaises(ConfigurationError):
+            utils.parse_resources(raw_resources)
+
+        # Source in other's destination.
+        raw_resources = """
+        /buckets/bid/collections/cid;/buckets/bid/collections/cid2;/buckets/bid/collections/cid3
+        /buckets/bid/collections/cid1;/buckets/bid/collections/cid2;/buckets/bid/collections/cid
```
Comment (glasserc) [inline tests/test_utils.py]: This test also repeats a preview.

File: kinto_signer/utils.py (hunk 4 above)
Comment (leplatrem) [reply kinto_signer/utils.py]: because we already raise with `if key in resources:`above

**Pull request merged**